File: 04-Linked-List/Test/src/LinkedList.java
/**
 * 链表
 */
public class LinkedList<E> {

    private class Node {
        public E e;
        public Node next;

        public Node(E e,Node next){
            this.e = e;
            this.next = next;
        }
    }

    private Node dummyHead;
    private int size;

    public LinkedList(){
        dummyHead = new Node(null,null);
        size = 0;
    }

    public int getSize(){
        return size;
    }

    public boolean isEmpty(){
        return size == 0;
    }

    // 添加元素
    public void add(int index, E e){
        if (index < 0 || index > size)
            throw new IllegalArgumentException("index Illegal");

        Node prev = dummyHead;
        // 跳过index之前的节点
        for (int i = 0; i < index; i++)
            prev = prev.next;

        prev.next = new Node(e,prev.next);
        size++;
    }

    // 头部添加元素
    public void addFirst(E e){
        add(0,e);
    }

    // 末尾添加元素
    public void addLast(E e){
        add(size,e);
    }

    // 返回 index 索引的元素值
    public E get(int index){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("index Illegal");

        Node cur = dummyHead.next;
        for (int i = 0; i < index; i ++)
            cur = cur.next;

        return cur.e;
    }

    public E getFirst(){
        return get(0);
    }

    public E getLast(){
        return get(size-1);
    }

    // 更新index的值
    public void set(int index, E e){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("Set Failed. Illegal index.");

        Node cur = dummyHead.next;
        for (int i = 0; i < index; i++)
            cur = cur.next;
        
        cur.e = e;
    }
    
    
    // 是否包含元素 e
    public boolean contains(E e){
        Node cur = dummyHead.next;

        while (cur != null) {
            if (cur.e.equals(e))
                return true;
            cur = cur.next;
        }

        return false;
    }

    // 删除元素
    public E remove(int index){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("index Illegal");

        Node prev = dummyHead;
        for (int i = 0; i < index; i ++)
            prev = prev.next;

        Node retNode = prev.next;
        prev.next = retNode.next;
        retNode.next = null;
        size--;

        return retNode.e;
    }

    // 删除头节点
    public E removeFirst(){
        return remove(0);
    }

    // 删除尾节点
    public E removeLast(){
        return remove(size-1);
    }
}

File: go/04-Linked-List/04-Query-and-Update-in-LinkedList/linkedlist/linkedlist.go
package linkedlist

import (
	"bytes"
	"fmt"
	"github.com/dingsw2019/play-with-data-struct/go/utils"
)

type Node struct {
	e    interface{}
	next *Node
}

func (n *Node) String() string {
	return fmt.Sprint(n.e)
}

type LinkedList struct {
	dummyHead *Node // 虚拟头节点, 不计入 size
	size      int
}

func New() *LinkedList {
	return &LinkedList{
		dummyHead: &Node{},
	}
}

// 获取链表中的元素个数
func (l *LinkedList) GetSize() int {
	return l.size
}

// 返回链表是否为空
func (l *LinkedList) IsEmpty() bool {
	return l.size == 0
}

// 添加元素
func (l *LinkedList) Add(index int, e interface{}) {
	if index < 0 || index > l.size {
		panic("add failed, index out of range")
	}
	prev := l.dummyHead
	for i := 0; i < index; i++ {
		prev = prev.next
	}
	prev.next = &Node{e, prev.next}
	l.size++
}
func (l *LinkedList) AddFirst(e interface{}) {
	l.Add(0, e)
}
func (l *LinkedList) AddLast(e interface{}) {
	l.Add(l.size, e)
}

// 获取元素
func (l *LinkedList) Get(index int) interface{} {
	if index < 0 || index >= l.size {
		panic("get failed, index out of range")
	}
	cur := l.dummyHead.next
	for i := 0; i < index; i++ {
		cur = cur.next
	}
	return cur.e
}
func (l *LinkedList) GetFirst() interface{} {
	return l.Get(0)
}
func (l *LinkedList) GetLast() interface{} {
	return l.Get(l.size)
}

func (l *LinkedList) Set(index int, e interface{}) {
	if index < 0 || index >= l.size {
		panic("set failed, index out of range")
	}
	cur := l.dummyHead.next
	for i := 0; i < index; i++ {
		cur = cur.next
	}
	cur.e = e
}

func (l *LinkedList) Contains(e interface{}) bool {
	cur := l.dummyHead.next
	for cur != nil {
		if utils.Compare(cur.e, e) == 0 {
			return true
		}
		cur = cur.next
	}
	return false
}

func (l *LinkedList) Remove(index int) interface{} {
	if index < 0 || index >= l.size {
		panic("remove failed, index out of range")
	}
	prev := l.dummyHead
	for i := 0; i < index; i++ {
		prev = prev.next
	}
	node := prev.next
	prev.next = node.next
	node.next = nil
	l.size--
	return node.e
}
func (l *LinkedList) RemoveFirst() interface{} {
	return l.Remove(0)
}
func (l *LinkedList) RemoveLast() interface{} {
	return l.Remove(l.size - 1)
}

func (l *LinkedList) String() string {
	var buf bytes.Buffer
	cur := l.dummyHead.next
	for cur != nil {
		buf.WriteString(fmt.Sprint(cur.e) + "->")
		cur = cur.next
	}
	buf.WriteString("NULL")
	return buf.String()
}
